src/main/java/uk/gov/hmcts/reform/blobrouter/util/ZipVerifiers.java

```diff
@@ -152,23 +155,52 @@ private static PublicKey decodePublicKey(String publicKeyBase64) {
         }
     }
 
+    public static class PublicKeyFile {
+        public final String derFilename;
+        public final String publicKeyBase64;
+
+        public PublicKeyFile(String derFilename, String publicKeyBase64) {
+            this.derFilename = derFilename;
+            this.publicKeyBase64 = publicKeyBase64;
+        }
+    }
+
     public static class ZipStreamWithSignature {
         public final ZipInputStream zipInputStream;
         public final String publicKeyBase64;
-        public final String zipFileName;
-        public final String container;
+
+        private static PublicKeyFile cachedPublicKeyFile;
 
         public ZipStreamWithSignature(
             ZipInputStream zipInputStream,
-            String publicKeyBase64,
-            String zipFileName,
-            String container
+            String publicKeyBase64
         ) {
             this.zipInputStream = zipInputStream;
             this.publicKeyBase64 = publicKeyBase64;
-            this.zipFileName = zipFileName;
-            this.container = container;
         }
 
+        public static ZipStreamWithSignature fromKeyfile(
+            ZipInputStream zipInputStream,
+            String publicKeyDerFile
+        ) {
+            try {
+                String publicKeyBase64 = null;
+                if (!Strings.isNullOrEmpty(publicKeyDerFile) && !"none".equalsIgnoreCase(publicKeyDerFile)) {
+                    if (cachedPublicKeyFile != null && publicKeyDerFile.equals(cachedPublicKeyFile.derFilename)) {
+                        publicKeyBase64 = cachedPublicKeyFile.publicKeyBase64;
+                    } else {
+                        publicKeyBase64 =
+                            Base64.getEncoder().encodeToString(
+                                Resources.toByteArray(getResource(publicKeyDerFile))
+                            );
+                        cachedPublicKeyFile = new PublicKeyFile(publicKeyDerFile, publicKeyBase64);
+                    }
+                }
+
+                return new ZipStreamWithSignature(zipInputStream, publicKeyBase64);
+            } catch (IOException e) {
+                throw new SignatureValidationException(e);
+            }
```

Comment:
This looks quite complex and imho it's not trivial to see what's going on here and when.
Perhaps it would be worth adding a test for this method?

Reply:
Added tests